### PR TITLE
 Added a status update on fwupd on FreeBSD 

### PR DESCRIPTION
**Pull request by mkopec**
Add a progress update for the FreeBSD port of fwupd.

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

### CODE REVIEW

File: blog/content/post/2021-06-14-fwupd-freebsd-status.md [hunk 1]
```diff
@@ -0,0 +1,1497 @@
...
+[previous](https://blog.3mdeb.com/2021/2021-02-16-fwupd-compilation-under-freebsd/)
+[entries](https://blog.3mdeb.com/2021/2021-02-16-fwupd-compilation-under-freebsd/),
+I encourage you to do so.
+
+---
+
+## Introduction
+
+In this blog post, I will document how we created a working port of fwupd for
+FreeBSD as well as the biggest challenges we encountered.
```
Comment (Asiderr) [inline blog/content/post/2021-06-14-fwupd-freebsd-status.md:31]:
```suggestion
FreeBSD, as well as the biggest challenges we encountered.
```

File: blog/content/post/2021-06-14-fwupd-freebsd-status.md [hunk 2]
```diff
@@ -0,0 +1,1497 @@
...
+We found we had to implement 3 basic functionalities in order to have a basic
+FreeBSD port:
+
+- Gathering information about applicable updates for attached devices,
+- Updating firmware of devices attached via USB,
+- Updating the system firmware via the UEFI Update Capsule mechanism.
+
+## Identifying available updates
+
+One difference between the Linux and FreeBSD APIs is the lack of `memfd_create(),`
```
Comment (Asiderr) [inline blog/content/post/2021-06-14-fwupd-freebsd-status.md:42]: Not true, it was added in FreeBSD 13.0, but we started implementation for version 12.2, where it is not included.

File: blog/content/post/2021-06-14-fwupd-freebsd-status.md [hunk 3]
```diff
@@ -0,0 +1,1497 @@
...
+
+        • Firmware updates to address the Intel Technical Advisory  .
+
+10:14:01:0389 FuPlugin             destroy(uefi_capsule)
+```
+
+## Applying USB updates
+
+Next, we encountered a problem with USB device updates: After a seemingly
+successful update of a device, - in our testing, a ColorHug2 - in would not
```
Comment (Asiderr) [inline blog/content/post/2021-06-14-fwupd-freebsd-status.md:369]: After reboot to the bootloader mode a device ...

Comment (Asiderr) [inline blog/content/post/2021-06-14-fwupd-freebsd-status.md:369]:
```
- in would not
+ didn't
```

File: blog/content/post/2021-06-14-fwupd-freebsd-status.md [hunk 4]
```diff
@@ -0,0 +1,1497 @@
...
+
+fwupd uses the libgusb library, a GLib wrapper around libusb. The usual flow
+of an update is as follows:
+
+1. Issue command to device to enter bootloader mode - in the case of ColorHug2,
+a custom HID-based flashing protocol
+2. Write update to device
+3. Upon successful update, return device to runtime mode
+
+The issue occured after the last step - we were unable to reattach the device to
```
Comment (Asiderr) [inline blog/content/post/2021-06-14-fwupd-freebsd-status.md:380]: Not true, after the first step.

File: blog/content/post/2021-06-14-fwupd-freebsd-status.md [hunk 5]
```diff
@@ -0,0 +1,1497 @@
...
+
+From the perspective of security, UEFI updates are absolutely critical and
+implementing this functionality for FreeBSD was a priority. There were a couple
+of parts that would need to be implemented in order to make it work:
+
+- UEFI ESRT table support in FreeBSD, and support for it in fwupd
+- FreeBSD efivar backend for fwupd - on Linux, efivar support is implemented
+via a `sysfs` interface, while FreeBSD has a C API
+- `bsdisks` support in fwupd for EFI system partition detection
+
```
Comment (Asiderr) [inline blog/content/post/2021-06-14-fwupd-freebsd-status.md:936]:
```suggestion
- adding support in a `fwupd` daemon plugin - the UEFI update capsule plugin 
```

File: blog/content/post/2021-06-14-fwupd-freebsd-status.md [hunk 6]
```diff
@@ -0,0 +1,1497 @@
...
+UEFI ESRT (EFI System Resource Table) is a standard interface for firmware
+updates available since UEFI 2.5. It exposes, among other data, information
+about the currently installed firmware versions and the status of last
+update attempt. It's used by fwupd for detection and matching availabile
+updates. Support for these tables was missing in FreeBSD - so we added it.
+Upstream patch available [here](https://github.com/3mdeb/freebsd-src/pull/2).
```
Comment (Asiderr) [inline blog/content/post/2021-06-14-fwupd-freebsd-status.md:942]: Link official upstream review https://reviews.freebsd.org/D30104

File: blog/content/post/2021-06-14-fwupd-freebsd-status.md [hunk 7]
```diff
@@ -0,0 +1,1497 @@
...
+fwupd applies firmware updates by installing a small EFI binary along with the
+update capsule into the ESP and setting the EFI bootnext variable to point to it.
+The machine reboots and launches the EFI binary which then calls
+`UpdateCapsule()`, which in turn tells the UEFI to apply the capsule. The
+actual flashing is handled by the UEFI implementation itself.
+
+This requires efivar support, and FreeBSD has a different, programmatic
+API, so support for it had to be added in fwupd. Furthermore, since FreeBSD has
+a disk management API that differs slightly from the Linux standard UDisks2 API,
+support for it also had to be added. Patches available
```
Comment (Asiderr) [inline blog/content/post/2021-06-14-fwupd-freebsd-status.md:953]: Patches are currently in the upstream.

**Review by Asiderr — CHANGES_REQUESTED** [review]

File: blog/content/post/2021-06-14-fwupd-freebsd-status.md [hunk 8]
```diff
@@ -0,0 +1,1500 @@
...
+We found we had to implement 3 basic functionalities in order to have a basic
+FreeBSD port:
+
+- Gathering information about applicable updates for attached devices,
+- Updating firmware of devices attached via USB,
+- Updating the system firmware via the UEFI Update Capsule mechanism.
+
+## Identifying available updates
+
+One difference between Linux and FreeBSD 12.2, which we started development
```
Comment (Asiderr) [inline blog/content/post/2021-06-14-fwupd-freebsd-status.md:42]:
```suggestion
One difference between Linux and FreeBSD 12.2, on which we started development
```

File: blog/content/post/2021-06-14-fwupd-freebsd-status.md [hunk 9]
```diff
@@ -0,0 +1,1500 @@
...
+on, is the lack of `memfd_create(),` a function that creates an anonymous,
```
Comment (Asiderr) [inline blog/content/post/2021-06-14-fwupd-freebsd-status.md:43]:
```suggestion
is the lack of `memfd_create(),` a function that creates an anonymous,
```

File: blog/content/post/2021-06-14-fwupd-freebsd-status.md [hunk 10]
```diff
@@ -0,0 +1,1500 @@
...
+
+        • Firmware updates to address the Intel Technical Advisory  .
+
+10:14:01:0389 FuPlugin             destroy(uefi_capsule)
+```
+
+## Applying USB updates
+
+Next, we encountered a problem with USB device updates: When we tried to update
+the firmware of a ColorHug2, after rebooting it to the device's bootloader mode,
```
Comment (Asiderr) [inline blog/content/post/2021-06-14-fwupd-freebsd-status.md:371]:
```suggestion
the firmware of a ColorHug2, and after rebooting it to the device's bootloader mode,
```

**Review by Asiderr — APPROVED** [review]

File: blog/content/post/2021-06-14-fwupd-freebsd-status.md [hunk 11]
```diff
@@ -0,0 +1,1499 @@
+---
+title: fwupd on FreeBSD - Status Update
+abstract: 'An update on the current status of the BSD port of fwupd.'
```
Comment (artur-rs) [inline blog/content/post/2021-06-14-fwupd-freebsd-status.md:3]: @mkopec @Asiderr it would be more appealing to have abstract with at least 2-3 sentences

**Comment by artur-rs** [comment]
@mkopec please run the spellchecker, e.g. [Grammarly](https://app.grammarly.com/)

**Comment by artur-rs** [comment]
@mkopec @Asiderr those logs are quite big and it't not easy to scroll horizontally. Have you thought about ascinema recording?

**Review by artur-rs — APPROVED** [review]

**Pull request merged**